paddle/fluid/framework/ir/op_fusion_pass.cc

```diff
@@ -0,0 +1,357 @@
+/* Copyright (c) 2018 PaddlePaddle Authors. All Rights Reserved.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License. */
+
+#include <algorithm>
+#include <unordered_set>
+
+#include "paddle/fluid/framework/ir/op_fusion_pass.h"
+#include "paddle/fluid/framework/op_proto_maker.h"
+#include "paddle/fluid/framework/operator.h"
+
+namespace paddle {
+namespace framework {
+namespace ir {
+const char kOpDescs[] = "op_descs";
+typedef std::vector<std::unique_ptr<OpDesc>> OpDescs;
+
+std::unique_ptr<ir::Graph> OpFusionPass::ApplyImpl(
+    std::unique_ptr<ir::Graph> graph) const {
+  graph->Set<OpDescs>(kOpDescs, new OpDescs);
+
+  std::vector<ir::Node *> topo_order = ir::TopologySortOperations(*graph.get());
+
+  std::unordered_map<const Node *, Node *> internal_nodes;
+  std::unordered_set<ir::Node *> need_removed_nodes;
+
+  for (auto iter_node = topo_order.rbegin(); iter_node != topo_order.rend();
+       ++iter_node) {
+    auto cur_node = *iter_node;
+
+    std::unordered_set<Node *> tobe_fused;
+
+    if (SetupFusion(cur_node, internal_nodes, &tobe_fused)) {
+      Node *new_node =
+          FuseOperators(cur_node, tobe_fused, &need_removed_nodes, graph.get());
+
+      tobe_fused.emplace(cur_node);
+      need_removed_nodes.insert(tobe_fused.begin(), tobe_fused.end());
+      for (auto &sub_node : tobe_fused) {
+        PADDLE_ENFORCE_EQ(internal_nodes.count(sub_node), 0);
+        internal_nodes.emplace(sub_node, new_node);
+      }
+    }
+  }
+
+  if (VLOG_IS_ON(6)) {
+    PrintTopologySort(topo_order, "TopologySortOperations(before fusing): ");
+  }
+
+  // Release unnecessary node
+  for (auto &node : need_removed_nodes) {
+    graph->ReleaseNode(node);
+  }
+
+  if (VLOG_IS_ON(6)) {
+    std::vector<ir::Node *> topo_order_after =
+        ir::TopologySortOperations(*graph.get());
+    PrintTopologySort(topo_order_after,
+                      "TopologySortOperations(after fusing): ");
+  }
+  return graph;
+}
+
+bool OpFusionPass::SetupFusion(
+    const NodePtr node,
+    const std::unordered_map<const Node *, InternalNodePtr> &internal_nodes,
+    std::unordered_set<Node *> *tobe_fused) const {
+  PADDLE_ENFORCE(!node->IsVariable(), "Node should not be variable.");
```

Comment:
`PADDLE_ENFORCE(node->IsOp)` ?

Why we need IsVariable() method. it does not make sense since the enum can be changed time by time.